### PR TITLE
gpio: pcal64xxa: switch to 8 bit internal data automatically

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Convert the internal uint16_t data of the driver to an internal type that automatically switches from uint8_t to uint16_t depending on whether any 16 bit device is present in the system or not. This shrinks the internal structures by few bytes when the extra data is not needed.

### CODE REVIEW

**Comment by fabiobaltieri** [comment]
@bbilas I think you were right, I do enjoy optimizing drivers after all

**Review by benediktibk — APPROVED** [review]
LGTM

**Comment by fabiobaltieri** [comment]
Fixed wrong type for `PCAL64XXA_INIT_HIGH` (should have been unsgined)

**Review by cfriedt — DISMISSED** [review]

File: drivers/gpio/gpio_pcal64xxa.c [hunk 1]
```diff
@@ -54,17 +54,29 @@ enum pcal6416a_register {
 	PCAL6416A_REG_OUTPUT_PORT_CONFIGURATION = 0x4F,
 };
 
+#if DT_HAS_COMPAT_STATUS_OKAY(nxp_pcal6416a)
+typedef uint16_t pcal64xxa_data_t;
+#define PCAL64XXA_INIT_HIGH UINT16_MAX
+#define PRIpcal_data "%04X"
```
Comment (gmarull) [inline drivers/gpio/gpio_pcal64xxa.c:60]:
```suggestion
#define PRIpcal_data "04” PRIx16
```

File: drivers/gpio/gpio_pcal64xxa.c [hunk 2]
```diff
@@ -54,17 +54,29 @@ enum pcal6416a_register {
...
+#elif DT_HAS_COMPAT_STATUS_OKAY(nxp_pcal6408a)
+typedef uint8_t pcal64xxa_data_t;
+#define PCAL64XXA_INIT_HIGH UINT8_MAX
+#define PRIpcal_data "%02X"
```
Comment (gmarull) [inline drivers/gpio/gpio_pcal64xxa.c:64]:
```suggestion
#define PRIpcal_data "02" PRIx8
```

File: drivers/gpio/gpio_pcal64xxa.c [hunk 3]
```diff
@@ -271,16 +284,16 @@ static int pcal64xxa_port_get_raw(const struct device *dev, gpio_port_value_t *v
...
 {
 	const struct pcal64xxa_drv_cfg *drv_cfg = dev->config;
 	struct pcal64xxa_drv_data *drv_data = dev->data;
 	int rc;
-	uint16_t output;
+	pcal64xxa_data_t output;
 
-	LOG_DBG("setting port with mask 0x%04X with value 0x%04X and toggle 0x%04X", mask, value,
-		toggle);
+	LOG_DBG("setting port with mask 0x" PRIpcal_data " with value 0x" PRIpcal_data
```
Comment (gmarull) [inline drivers/gpio/gpio_pcal64xxa.c:295]:
```suggestion
	LOG_DBG("setting port with mask 0x" PRIpcal_data " with value 0x%" PRIpcal_data
```

File: drivers/gpio/gpio_pcal64xxa.c [hunk 4]
```diff
@@ -271,16 +284,16 @@ static int pcal64xxa_port_get_raw(const struct device *dev, gpio_port_value_t *v
...
+		" and toggle 0x" PRIpcal_data, mask, value, toggle);
```
Comment (gmarull) [inline drivers/gpio/gpio_pcal64xxa.c:296]:
```suggestion
		" and toggle 0x%" PRIpcal_data, mask, value, toggle);
```

File: drivers/gpio/gpio_pcal64xxa.c (hunk 1 above)
Comment (fabiobaltieri) [reply drivers/gpio/gpio_pcal64xxa.c]: done

File: drivers/gpio/gpio_pcal64xxa.c (hunk 2 above)
Comment (fabiobaltieri) [reply drivers/gpio/gpio_pcal64xxa.c]: done

File: drivers/gpio/gpio_pcal64xxa.c (hunk 3 above)
Comment (fabiobaltieri) [reply drivers/gpio/gpio_pcal64xxa.c]: done

File: drivers/gpio/gpio_pcal64xxa.c (hunk 4 above)
Comment (fabiobaltieri) [reply drivers/gpio/gpio_pcal64xxa.c]: and done, you can tell I don't use these often :-) did the same on the gpio-kbd driver, will adjust that as well

**Review by gmarull — APPROVED** [review]

**Review by benediktibk — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**